packages/hub/src/graphql/queries/search.ts

```diff
@@ -0,0 +1,126 @@
+            setweight(to_tsvector(coalesce("SquiggleSnippet".code, '')), 'B') ||
+            setweight(to_tsvector(coalesce("RelativeValuesDefinition".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("UserOwner".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("GroupOwner".slug, '')), 'A'),
+            websearch_to_tsquery(${text})
+          ) AS rank,
+          ts_headline(
+            concat_ws(
+              ' ',
+              "Model".slug,
+              "RelativeValuesDefinition".slug,
+              "UserOwner".slug,
+              "GroupOwner".slug
+            ),
+            websearch_to_tsquery(${text}),
+            'HighlightAll=t'
+          ) AS "slugSnippet",
+          ts_headline(concat_ws(
+              ' ',
+              "SquiggleSnippet".code
+            ), websearch_to_tsquery(${text})
+          ) AS "textSnippet"
+        FROM "Searchable"
+        LEFT JOIN "Model" ON "Model".id = "Searchable"."modelId"
+          LEFT JOIN "Owner" AS "ModelOwner" ON "Model"."ownerId" = "ModelOwner".id
+            LEFT JOIN "User" AS "ModelOwnerUser" ON "ModelOwner".id = "ModelOwnerUser"."ownerId"
+            LEFT JOIN "Group" AS "ModelOwnerGroup" ON "ModelOwner".id = "ModelOwnerGroup"."ownerId"
+          LEFT JOIN "ModelRevision" ON "Model"."currentRevisionId" = "ModelRevision".id
+            LEFT JOIN "SquiggleSnippet" ON "ModelRevision"."contentId" = "SquiggleSnippet".id
+        LEFT JOIN "RelativeValuesDefinition" ON "RelativeValuesDefinition".id = "Searchable"."definitionId"
+          -- LEFT JOIN "Owner" AS "RelativeValuesDefinitionOwner" ON "RelativeValuesDefinition"."ownerId" = "RelativeValuesDefinitionOwner".id
+        LEFT JOIN "User" ON "User".id = "Searchable"."userId"
+          LEFT JOIN "Owner" AS "UserOwner" ON "User"."ownerId" = "UserOwner".id
+        LEFT JOIN "Group" ON "Group".id = "Searchable"."groupId"
+          LEFT JOIN "Owner" AS "GroupOwner" ON "Group"."ownerId" = "GroupOwner".id
+        WHERE
+          (
+            -- check permissions, should match "modelWhereHasAccess" function
+            "Model".id IS NULL OR
+            "Model"."isPrivate" IS FALSE
+            OR "ModelOwnerUser".email = ${session?.user.email}
+            OR "ModelOwnerGroup".id IN (
+              SELECT DISTINCT "groupId"
+              FROM "UserGroupMembership"
+              LEFT JOIN "User" ON "User".id = "UserGroupMembership"."userId"
+              WHERE "User".email = ${session?.user.email}
+            )
+          ) AND (
+            -- construct search document
+            to_tsvector(concat_ws(
+              ' ',
+              "Model".slug,
+              "SquiggleSnippet".code,
+              "RelativeValuesDefinition".slug,
+              "UserOwner".slug,
+              "GroupOwner".slug
+            )) @@ websearch_to_tsquery(${text})
+          )
+        ORDER BY rank DESC
+        LIMIT ${limit}
+        OFFSET ${offset}
```

Comment:
I could add more comments to this query, but I don't think it'll be more useful than just skimming the first few chapters of https://www.postgresql.org/docs/current/textsearch.html.